src/avl.py

```diff
@@ -0,0 +1,468 @@
+        self.value = None
+        self._left = None
+        self._right = None
+        self.parent = None
+        self.container = container
+        self.left_height = self.right_height = 0
+
+    @property
+    def left(self):
+        return self._left
+
+    @property
+    def right(self):
+        return self._right
+
+    def __str__(self, depth=0):
+        def helper(child):
+            """Simplifies printing children."""
+
+            return "\n" + (child.__str__(depth + 1)
+                           if child is not None
+                           else "  " * (depth + 1) + "None")
+
+        return ("  " * depth + (
+            "AVLTreeNode("
+            "id = %s, "
+            "value = %s, "
+            "balance_factor = %s, "
+            "left_height = %s, "
+            "right_height = %s)"
+        ) % (id(self),
+             self.value,
+             self.balance_factor,
+             self.left_height,
+             self.right_height
+            )) + helper(self.left) + helper(self.right)
+
+    @property
+    def max_height(self) -> int:
+        """Returns the height of the maximal depth subtree."""
+
+        return max(self.left_height, self.right_height)
+
+    @property
+    def leaf(self) -> bool:
+        """Returns whether this node is a leaf or not."""
+
+        return self.left is None and self.right is None
+
+    def become_branch(self):
+        """Converts this node from a leaf to a branch."""
+
+        self._left = AVLTreeNode(self.container)
+        self._right = AVLTreeNode(self.container)
+        self._left.parent = self._right.parent = self
+        self.fix_height_metadata()
+
+    @property
+    def balance_factor(self) -> int:
+        """This represents how, and in which direction, this node is imbalanced."""
+
+        return self.left_height - self.right_height
+
+    def update_log(self, operation, force=False):
+        """Updates the log of operation and tree pairs for later visualization.
+
+        The log is only ever updated at the root, to avoid visualizing multiple layers
+        in one stream, which would be very confusing.
+
+        Args:
+          operation: A textual description of the operation taking place, intended
+            for human consumption.
+          force: Defaults to false. If this is true, we will append to the log regardless
+            of whether or not we are the root. This is used during operations where the
+            root has just changed, but the whole tree should still be visualized.
+        """
+
+        # if we aren't the root, don't log
+        if not force and (self.container.root != self):
+            return
+
+        self.container.log.append([operation, self.container.root.to_dict()])
+
+    def insert(self, value):
+        """Inserts the given value into the appropriate spot in the tree.
+
+        Args:
+          value: the value to insert.
+        """
+
+        force = self.container._root == self
```

Comment:
I'm honestly not sure, but it's worth investigating. This code got pretty messy for a while.